test/vi-mapping-test-app/src/vi-mapping-test-app.cc

```diff
@@ -235,6 +236,56 @@ void VIMappingTestApp::corruptCameraExtrinsics(
   }
 }
 
+void VIMappingTestApp::addCorruptDuplicateLandmarkObservations(int every_nth) {
+  vi_map::VIMapManager map_manager;
+  vi_map::VIMapManager::MapWriteAccess map =
+      map_manager.getMapWriteAccess(map_key_);
+
+  // Get set of first two vertices that shouldn't be corrupted.
+  vi_map::MissionIdList mission_ids;
+  map->getAllMissionIds(&mission_ids);
+  pose_graph::VertexIdSet not_to_corrupt_vertices;
+  vi_map::LandmarkIdSet landmark_ids;
+  map->getAllLandmarkIds(&landmark_ids);
+  int index = 0;
+  for (const vi_map::LandmarkId& landmark_id : landmark_ids) {
+    vi_map::Landmark& landmark = map->getLandmark(landmark_id);
+    const vi_map::KeypointIdentifierList& landmark_observations =
+        landmark.getObservations();
+    for (const auto& keypoint : landmark_observations) {
+      const pose_graph::VertexId& vertex_id = keypoint.frame_id.vertex_id;
+      if (index % every_nth != 0 || !keypoint.isValid() ||
+          !vertex_id.isValid()) {
+        ++index;
+        continue;
+      }
+
+      CHECK(map->hasVertex(vertex_id));
+      vi_map::Vertex& vertex = map->getVertex(vertex_id);
+      const unsigned int frame_idx =
+          static_cast<unsigned int>(keypoint.frame_id.frame_index);
+      aslam::VisualFrame& frame = vertex.getVisualFrame(frame_idx);
+
+      const Eigen::Vector2d observation_point =
+          frame.getKeypointMeasurement(keypoint.keypoint_index);
+      const Eigen::Vector2d noise(10, 10);
+      const Eigen::Vector2d corrupted_point = observation_point + noise;
+
+      Eigen::Matrix2Xd corrupted_keypoint(2, 1);
+      corrupted_keypoint.col(0) = observation_point + noise;
+      const unsigned int corrupted_keypoint_idx =
+          static_cast<unsigned int>(frame.getNumKeypointMeasurements());
+      aslam::insertAdditionalKeypointsToVisualFrame(
```

Comment:
This is a slightly dangerous function, since it's not used in other parts of the code I'm not sure how well it does what it's supposed to do. On top of that the vertex will have an issue because this vector https://github.com/ethz-asl/maplab/blob/618a0e0154d0c6e6cd84b5ac827140b5a9b84cb0/map-structure/vi-map/include/vi-map/vertex.h#L294 needs to be equal in size to the number of observed landmarks, which it won't be unless you resize it.

Reply:
What I would do instead is just some of the existing observations I would just assign to this landmark, and not try to create new ones

Reply:
It should be automatically resized by https://github.com/ethz-asl/maplab/blob/7875e745d7e8548e9d7220b4b70f349a4f5a309f/test/vi-mapping-test-app/src/vi-mapping-test-app.cc#L280, no?